lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,385 @@
+    elif not tarfile.is_tarfile(collection_path):
+        raise AnsibleError("The collection path specified '%s' is not a tarball, use 'ansible-galaxy collection "
+                           "build' to create a proper release artifact." % collection_path)
+
+    display.display("Publishing collection artifact '%s' to %s" % (to_text(collection_path), server))
+
+    url = "%s/api/v2/collections/" % server
+
+    data, content_type = _get_mime_data(collection_path)
+    headers = {
+        'Content-type': content_type,
+        'Content-length': len(data),
+    }
+    if key:
+        headers['Authorization'] = "Token %s" % key
+    validate_certs = not ignore_certs
+
+    try:
+        resp = json.load(open_url(url, data=data, headers=headers, method='POST', validate_certs=validate_certs))
+    except urllib_error.HTTPError as err:
+        try:
+            err_info = json.load(err)
+        except (AttributeError, ValueError):
+            err_info = {}
+
+        code = err_info.get('code', 'Unknown')
+        message = err_info.get('message', 'Unknown error returned by Galaxy server.')
+
+        raise AnsibleError("Error when publishing collection (HTTP Code: %d, Message: %s Code: %s)"
+                           % (err.code, message, code))
+
+    display.vvv("Collection has been pushed to the Galaxy server %s" % server)
+    import_uri = resp['task']
+    if wait:
+        _wait_import(import_uri, key, validate_certs)
+        display.display("Collection has been successfully published to the Galaxy server")
+    else:
+        display.display("Collection has been pushed to the Galaxy server, not waiting until import has completed "
+                        "due to --no-wait being set. Import task results can be found at %s" % import_uri)
+
+
+# TODO: Remove this once Python 2.6 is fully excised from the controller tests
+@contextmanager
+def _open_tarfile(path, mode='r'):
+    tfile = tarfile.open(path, mode=mode)
+    yield tfile
+    tfile.close()
+
+
+def _get_galaxy_yml(galaxy_yml_path):
+    mandatory_keys = frozenset(['namespace', 'name', 'version', 'authors', 'readme'])
+    optional_strings = ('description', 'repository', 'documentation', 'homepage', 'issues', 'license_file')
+    optional_lists = ('license', 'tags', 'authors')  # authors isn't optional but this will ensure it is list
+    optional_dicts = ('dependencies',)
+    all_keys = frozenset(list(mandatory_keys) + list(optional_strings) + list(optional_lists) + list(optional_dicts))
+
+    try:
+        with open(to_bytes(galaxy_yml_path), 'rb') as g_yaml:
+            galaxy_yml = yaml.safe_load(g_yaml)
+    except YAMLError as err:
+        raise AnsibleError("Failed to parse the galaxy.yml at '%s' with the following error:\n%s"
+                           % (to_native(galaxy_yml_path), to_native(err)))
+
+    set_keys = set(galaxy_yml.keys())
+    missing_keys = mandatory_keys.difference(set_keys)
+    if len(missing_keys) > 0:
```

Comment:
I think this may read better:
```suggestion
    if missing_keys:
```